### PR TITLE
HRQB 13 - Pipeline structures and CLI commands

### PR DESCRIPTION
**Pull request by ghukill**
### Purpose and background context

This PR fully realizes "pipelines" in this HRQBClient, with CLI commands to invoke and manage them.

Most of the building blocks were there with base classes `PandasPickleTask` and `QuickbaseUpsertTask`, but this completes the picture by extending functionality for base class `HRQBPipelineTask`, and wiring up CLI commands that invoke these special "pipeline" Tasks.

A related change was made during this work defining a required method `PandasPickleTask.get_dataframe(...)`.  This mirrors the `QUickbaseUpsertTask.get_records(...)` pattern, where it's preferable to have a "worker" method on a Luigi Task class, such that you can test its output directly without the complexities of the `.run()` method which will write to a Target.

In both cases, `run()` _can_ be overridden if the Task requires it, but most tasks will be doing one of the following, and the defult `run()` method is well suited for this:

- creating new dataframes from external sources (Extract Tasks)
- taking an input dataframe and modifying the data somehow (Transform Tasks)
- taking an input dataframe and writing to Quickbase (Load Tasks)

Apologies for the large commit, though the vast majority of changes are tests and test fixtures.  The most relevant new functionality can be seen here:

- [CLI commands](https://github.com/MITLibraries/hrqb-client/pull/14/files#diff-fc93c09dcc950b18155c7189cc4a5135797ceebe156990ee4b51a280075d9e0eR46-R119)
  - [method to init `HRQBPipelineTask` from CLI class path arguments](https://github.com/MITLibraries/hrqb-client/pull/14/commits/9bc158addd35f439ff8e5163e1074d71932cfdf3)
- [new base Task `HRQBPipelineTask` methods](https://github.com/MITLibraries/hrqb-client/pull/13/commits/35d646ab0b46f71aa45fc03141785d97fbdf7667#diff-d86e429c2a92dbd803698c92d15054d84cd0886541acd33a95f26ec5e48d39e6R155-R210)
- [creation of `PandasPickleTask.get_dataframe()` method](https://github.com/MITLibraries/hrqb-client/pull/13/commits/35d646ab0b46f71aa45fc03141785d97fbdf7667#diff-d86e429c2a92dbd803698c92d15054d84cd0886541acd33a95f26ec5e48d39e6R98-R107)
  - [example usage by an Extract Task](https://github.com/MITLibraries/hrqb-client/pull/13/commits/35d646ab0b46f71aa45fc03141785d97fbdf7667#diff-52a8ed9fa9fb3bfc1b3996f5eb619d05ffd78905a66250ae543c4054865eaefaR11-R18)

### How can a reviewer manually see the effects of these changes?

While there are not any real pipelines defined yet (these will populate `hrqb.tasks.*` when created), fixtures have been created for automated testing that are demonstrative of how they will be structured.

You may need some env vars, as they are considered required, but you can set fake/placeholder values:
```shell
WORKSPACE=dev
SENTRY_DSN=None
LUIGI_CONFIG_PATH=hrqb/luigi.cfg
QUICKBASE_API_TOKEN=abc123
QUICKBASE_APP_ID=def456
```

#### Structure of `tests.fixtures.tasks`

This fixtures folder structure mirrors what Tasks will look like in the `hrqb.tasks` module:
```
tests/fixtures/tasks
├── __init__.py
├── extract.py
    --> ExtractAnimalColors
    --> ExtractAnimalNames
├── load.py
    --> LoadAnimals
    --> LoadAnimalsDebug
├── pipelines.py
    --> Animals
    --> AnimalsDebug
└── transform.py
    --> PrepareAnimals
```

While this mirrors how Tasks will actually be defined in `hrqb.tasks`, it's not very helpful for an intuitive sense of how they wire to each other.

#### Inspect full, annotated pipeline code

The fixture file [`tests/fixtures/full_annotated_pipeline.py`](https://github.com/MITLibraries/hrqb-client/pull/14/files#diff-3f9d42c2c1e7210d5649c8832af840456cdff33fafffa655eba90be5880b0422) exists to hold a full, annotated pipeline.  This can be useful to see how Tasks are largely defined by:

- indicating what tasks they depend on
- providing a `get_dataframe()` or `get_records()` method as appropriate

The next section shows how to run this via the CLI, but it can also be interacted with directly in a python REPL:

```python
from tests.fixtures.full_annotated_pipeline import *

pipeline_task = AlphaNumeric()

print(pipeline_task.pipeline_as_ascii())
"""
├── INCOMPLETE: AlphaNumeric()
   ├── INCOMPLETE: CombineLettersAndNumbers(table_name=, pipeline=AlphaNumeric, stage=Transform)
      ├── INCOMPLETE: MultiplyNumbers(table_name=, pipeline=AlphaNumeric, stage=Transform)
         ├── INCOMPLETE: GenerateNumbers(table_name=, pipeline=AlphaNumeric, stage=Extract)
      ├── INCOMPLETE: GenerateLetters(table_name=, pipeline=AlphaNumeric, stage=Extract)
"""
```

#### Run CLI command

The ultimate goal of the changes from this PR are to support CLI level commands to invoke pipelines, which is now possible.

NOTE: without setting `TARGETS_DIRECTORY` env var, the default location will be the app's `output/` folder.

##### 1- Run `status` command for a tests fixture pipeline:
```shell
pipenv run hrqb --verbose \
pipeline \
-pm tests.fixtures.tasks.pipelines \
-p AnimalsDebug \
status
```

Note the output showing all tasks are `INCOMPLETE`.  This status output also gives a high level picture of the pipeline structure, showing all Tasks and how they rely on one another (where going down the tree, each Task relies on the Task above it).  This is the same output as `pipeline_task.pipeline_as_ascii()` from above.

```
├── INCOMPLETE: AnimalsDebug()
   ├── INCOMPLETE: LoadAnimalsDebug(pipeline=AnimalsDebug, stage=Load, table_name=Animals)
      ├── INCOMPLETE: PrepareAnimals(pipeline=AnimalsDebug, stage=Transform, table_name=Animals)
         ├── INCOMPLETE: ExtractAnimalColors(table_name=, pipeline=AnimalsDebug, stage=Extract)
         ├── INCOMPLETE: ExtractAnimalNames(table_name=, pipeline=AnimalsDebug, stage=Extract)
```

##### 2- Run `run` command for a tests fixture pipeline:
```shell
pipenv run hrqb --verbose \
pipeline \
-pm tests.fixtures.tasks.pipelines \
-p AnimalsDebug \
run
```

There should be a handful of outputs, and now the pipeline status should show the tasks as green `COMPLETE`.  Looking in the `output/` folder, you should also see artifacts from the pipeline run:

```
output
├── AnimalsDebug__Extract__ExtractAnimalColors.pickle
├── AnimalsDebug__Extract__ExtractAnimalNames.pickle
├── AnimalsDebug__Load__LoadAnimalsDebug.json
├── AnimalsDebug__Transform__PrepareAnimals.pickle
```

Note the filename structure of `<PipelineName>__<Stage>__<TaskName>.<Extension>`.  By prefixing files with the pipeline name, this allows multiple pipelines to reuse the same Task logic without the risk of getting data from another pipeline.  Within the _same_ pipeline, the Task outputs will be reused without re-running the Task (e.g. Data Warehouse  queries where perhaps multiple Tasks will reuse that data).

But, what if you _do_ want to reuse the data from another pipeline run?  This will be sorted in future PRs that define actual HRQB pipelines, if and when that need arises; could be as simple as a flag to omit the pipeline prefix effectively making it a Pipeline-agnostic Task.

If you immediately _re-run_ the pipeline, you should see some output like this, indicating that no Tasks needed to re-run, as their Targets (files) already existed:

```
...
===== Luigi Execution Summary =====

Scheduled 1 tasks of which:
* 1 complete ones were encountered:
    - 1 AnimalsDebug()

Did not run any tasks   <-------------------------------
This progress looks :) because there were no failed tasks or missing dependencies

===== Luigi Execution Summary =====
...
```

##### 2- Run `remove-data` command for a tests fixture pipeline:
```shell
pipenv run hrqb --verbose \
pipeline \
-pm tests.fixtures.tasks.pipelines \
-p AnimalsDebug \
remove-data
```

Lastly, can run a command that will analyze the pipeline and remove any Targets (files) that may exist for the pipeline.  This can also be accomplished as an addon step when running a pipeline with `--remove-data`.

Looking at `output/` folder now, all files should be gone.  To double confirm, you can run the `status` command again and note that all Tasks are now `INCOMPLETE` again.

### Includes new or updated dependencies?
YES

### Changes expectations for external applications?
YES

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/HRQB-13

### Developer
- [X] All new ENV is documented in README
- [X] All new ENV has been added to staging and production environments
- [X] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,9 +40,80 @@ QUICKBASE_APP_ID=# Quickbase App ID
...
+
+```text
+Usage: -c pipeline [OPTIONS] COMMAND [ARGS]...
+
+Options:
+  -p, --pipeline TEXT          Pipeline Task class name imported from
+                               configured pipeline module, e.g. 'MyPipeline'
+                               [required]
+  -pm, --pipeline-module TEXT  Module where Pipeline Task class is defined.
+                               Default: 'hrqb.tasks.pipelines'.
```
Comment (ehanson8) [inline README.md:71]: What's the scenario where the Pipeline Task class wouldn't be defined in `hrqb.tasks.pipelines`? Just wondering if this param is necessary

File: hrqb/base/task.py [hunk 2]
```diff
@@ -150,8 +164,69 @@ def run(self) -> None:
...
+    ) -> "HRQBPipelineTask":
+        """Factory method to import and instantiate an HRQBPipelineTask via class path."""
+        module = __import__(task_class_module, fromlist=[task_class_name])
+        task_class = getattr(module, task_class_name)
+        return task_class(**pipeline_parameters or {})
+
+    def pipeline_tasks_iter(
+        self, task: luigi.Task | None = None, level: int = 0
+    ) -> Iterator[tuple[int, luigi.Task]]:
+        """Yield Tasks, for all Tasks, in this Pipeline Task."""
```
Comment (ehanson8) [inline hrqb/base/task.py]: I am confident that the logic of all this is sound but this phrasing is hard to parse. Is there a clearer way to state this? I also think defining the params in the docstrings would also be helpful for the methods that have them. Some hand-holding for us `luigi` beginners would be appreciated 🙂

File: hrqb/base/task.py [hunk 3]
```diff
@@ -150,8 +164,69 @@ def run(self) -> None:
...
+    def pipeline_as_ascii(self) -> str:
+        """Return an ASCII representation of this Pipeline Task."""
+        output = ""
+        for level, task in self.pipeline_tasks_iter():
+            green_complete = "\033[92mCOMPLETE\033[0m"
+            red_incomplete = "\033[91mINCOMPLETE\033[0m"
+            status = green_complete if task.complete() else red_incomplete
+            indent = "   " * level + "├──"
+            output += f"\n{indent} {status}: {task}"
+        return output
```
Comment (ehanson8) [inline hrqb/base/task.py:236]: This is very helpful output

File: hrqb/cli.py [hunk 4]
```diff
@@ -39,3 +42,83 @@ def ping(ctx: click.Context) -> None:
...
+    "--pipeline-module",
+    type=str,
+    required=False,
+    help="Module where Pipeline Task class is defined. Default: 'hrqb.tasks.pipelines'.",
+    default="hrqb.tasks.pipelines",
+)
+@click.option(
+    "--pipeline-parameters",
+    callback=click_argument_to_dict,
+    help="Comma separated list luigi Parameters to pass to HRQBPipelineTask, "
```
Comment (ehanson8) [inline hrqb/cli.py]: `"Comma separated list of luigi Parameters...`

File: hrqb/cli.py [hunk 5]
```diff
@@ -39,3 +42,83 @@ def ping(ctx: click.Context) -> None:
...
+    pipeline_task = ctx.obj["PIPELINE_TASK"]
+    logger.warning("Removing all Pipeline Tasks Targets (data).")
+    logger.info(pipeline_task.remove_pipeline_targets())
+    logger.info("Successfully removed pipeline artifacts.")
+
+
+@pipeline.command()
+@click.option(
+    "--remove-data",
+    "remove_artifacts",  # renamed as variable to avoid collision with other command
```
Comment (ehanson8) [inline hrqb/cli.py]: What about just `--cleanup` which then invokes `remove_data` and avoids the collision-dodging rename?

File: hrqb/cli.py [hunk 6]
```diff
@@ -39,3 +42,83 @@ def ping(ctx: click.Context) -> None:
...
+    ctx: click.Context,
+    remove_artifacts: bool,  # noqa: FBT001
+) -> None:
+    pipeline_task = ctx.obj["PIPELINE_TASK"]
+    run_results = run_pipeline(pipeline_task)
+    message = f"Pipeline run result: {run_results.status.name}"
+    logger.info(message)
+    logger.info(pipeline_task.pipeline_as_ascii())
+    if remove_artifacts:
+        ctx.invoke(remove_data)
```
Comment (ehanson8) [inline hrqb/cli.py:123]: The logic of `status`, `run`, and `remove_data` is very simple and clear

File: tests/fixtures/full_annotated_pipeline.py
Comment (ehanson8) [inline tests/fixtures/full_annotated_pipeline.py:1]: Great addition, this is very, very helpful for getting my head wrapped around `luigi` workflows

File: tests/test_cli.py [hunk 7]
```diff
@@ -1,17 +1,136 @@
...
+from _pytest.logging import LogCaptureFixture
+from click.testing import Result
+
+from hrqb import cli
 
 OKAY_RESULT_CODE = 0
 MISSING_CLICK_ARG_RESULT_CODE = 2
 
 
+def text_in_logs_or_stdout(text, caplog: LogCaptureFixture, result: Result):
```
Comment (ehanson8) [inline tests/test_cli.py:10]: Clever method!

File: tests/test_pipelines.py
Comment (ehanson8) [inline tests/test_pipelines.py:1]: Very well-designed tests as always

**Review by ehanson8 — COMMENTED** [review]
Looking good, some questions

File: README.md (hunk 1 above)
Comment (ghukill) [reply README.md:71]: The CLI command in this PR uses that to run a pipeline defined in `tests/fixtures`:
```shell
pipenv run hrqb --verbose \
pipeline \
-pm tests.fixtures.tasks.pipelines \  #<--------------------
-p AnimalsDebug \
status
```

To your point, I don't see this getting used often for actual day-to-day pipelines for this application.  But it's a convention present in luigui with the `--module` CLI arg, and I think it adds some handy flexibility for testing, without much overhead.

File: hrqb/base/task.py (hunk 2 above)
Comment (ghukill) [reply hrqb/base/task.py]: You zeroed right in!  As I was re-reading the code after the PR, this caught me too.  Thanks for flagging.  I'll update the docstrings for these methods as needed for a bit more clarity.

File: hrqb/base/task.py (hunk 3 above)
Comment (ghukill) [reply hrqb/base/task.py:236]: Good to hear!  I thought so too.  For pretty minimal effort, can get a glance of the pipeline hierarchy _and_ status of the tasks.

File: hrqb/cli.py (hunk 5 above)
Comment (ghukill) [reply hrqb/cli.py]: I like this.  I was aiming for symmetry between the flag name and the actual CLI command it would run, but I think "cleanup" accurately reflects that it's something you do _after_ a run.

File: tests/fixtures/full_annotated_pipeline.py
Comment (ghukill) [reply tests/fixtures/full_annotated_pipeline.py:1]: Also glad to hear.  Took a gamble that, while defining a pipeline in a single file is unusual give the distinct E/T/L file structure that will actually be the case, a single file feels incredibly helpful for seeing a simple case in one place.

File: tests/test_cli.py (hunk 7 above)
Comment (ghukill) [reply tests/test_cli.py:10]: I did _not_ want to untangle why `pytest` was hot-swapping `result.output` and `caplog.text` when ultimately it wasn't much of an issue.  I think this method will also serve as a reminder if we do find that logs are scrambled at all in Cloudwatch, that here could lie the issue.

**Comment by ghukill** [comment]
Thanks for the thoughtful review @ehanson8, particularly given the size of changes.  Updates in new commit.

File: hrqb/base/task.py [hunk 8]
```diff
@@ -199,15 +199,22 @@ def init_task_from_class_path(
     def pipeline_tasks_iter(
         self, task: luigi.Task | None = None, level: int = 0
     ) -> Iterator[tuple[int, luigi.Task]]:
-        """Yield Tasks, for all Tasks, in this Pipeline Task."""
+        """Yield all Tasks that are part of the dependency chain for this Task.
+
+        This method begins with the Pipeline Task itself, then recursively discovers and
+        yields parent Tasks as they are required.
+        """
```
Comment (ehanson8) [inline hrqb/base/task.py:211]: So much clearer, thank you!

**Review by ehanson8 — APPROVED** [review]
Looks great, thanks for the changes!

File: hrqb/base/task.py [hunk 9]
```diff
@@ -1,17 +1,20 @@
...
+import logging
 import os
 from abc import abstractmethod
+from collections.abc import Iterator
 from typing import Literal
 
 import luigi  # type: ignore[import-untyped]
 import pandas as pd
 
 from hrqb.base import PandasPickleTarget, QuickbaseTableTarget
```
Comment (jonavellecuerdo) [inline hrqb/base/task.py:12]: Can the `target` attribute be converted to a property on these `HRQBTask` classes using `@property`? 🤔

**Review by jonavellecuerdo — COMMENTED** [review]
@ghukill Lots of cool methods implemented here! I agree with @ehanson8 that the `HRQBPipelineTask.pipeline_as_ascii()` is a really helpful feature. Just had one question / minor suggested change.

File: hrqb/base/task.py (hunk 9 above)
Comment (ghukill) [reply hrqb/base/task.py:12]: Am I understanding that you'd want to be able to use `self.target` instead of `self.target()` in places like this?

```python
class PandasPickleTask(HRQBTask):    

    def run(self) -> None:
        self.target.write(self.get_dataframe())   # <----------------?
```

If so, I think it'd get tricky.  `target()` is a luigi method from `luigi.Task` that `HRQBTask` extends.  Actually, one reason these classes exist is so that actual downstream tasks won't have to interact with `self.target()` very often.  They should like either define `get_dataframe()` or `get_records()`, which themselves interact with `self.target()`. 

In short: `target` is a reserved method from `luigi.Task`, so could be hard to `@property`-ify.

Comment (ghukill) [reply hrqb/base/task.py:12]: Thanks for the conversation on slack.  During that, realizing that I misspoke above and `target` is _not_ a reserved property or method on `luigi.Task`.  As discussed, I think this was me early on in the project thinking it'd be handy to refer directly to a Task's "target" as an object to receive and work with vs what feels like a more opaque `output()` method which just so happens to return a target class.

Going to dig in a bit and see if some renaming or reworking might make this a little cleaner.  Thanks for raising!

Comment (ghukill) [reply hrqb/base/task.py:12]: @jonavellecuerdo - this commit was added that addresses your original request / question: https://github.com/MITLibraries/hrqb-client/pull/14/commits/142cfb80f0727d8cf5fe96ec5e54809bcd492f4a.

Overall, I think it's a good one, thanks for raising.

To summarize the commit message a bit: the goal of `HRQBTask.target()` was to more accurately reflect that it's returning an actual `luigi.Target` instance, where the default requirement in luigi is to define a `<Task>.output()` method that does the same.  All in all, it's mostly just a project-specific preference to use the word "target" (which feels more accurate to what it returns) over the default luigi "output".  But by making it a property, as you asked about, I think it further leans into this opinionation in a good way!  

The end result is the ability to do `MyTask.target` and get the fully instantiated target instead of `MyTask.output()` which, to me, feels a bit vague as to what it's returning or doing.

**Review by jonavellecuerdo — APPROVED** [review]
Thank you for making these updates!
